### PR TITLE
Use BFS instead of Full Scan to browse Gdrive folders

### PR DESCRIPTION
**Pull request by lasryaric**

### CODE REVIEW

File: connectors/src/connectors/google_drive/temporal/workflows.ts [hunk 1]
```diff
@@ -53,26 +54,36 @@ export async function googleDriveFullSync(
...
-        nextPageToken
-      );
-      nextPageToken = res.nextPageToken ? res.nextPageToken : undefined;
-      totalCount += res.count;
-      await reportInitialSyncProgress(
-        connectorId,
-        `Synced ${totalCount} files`
-      );
-    } while (nextPageToken);
+      do {
```
Comment (spolu) [inline connectors/src/connectors/google_drive/temporal/workflows.ts:63]: The signal should interrupt the outer while loop, not this one?

Comment (spolu) [reply connectors/src/connectors/google_drive/temporal/workflows.ts:63]: Should probably interrupt both really

File: connectors/src/connectors/google_drive/temporal/workflows.ts [hunk 2]
```diff
@@ -53,26 +54,36 @@ export async function googleDriveFullSync(
...
     let nextPageToken: string | undefined = undefined;
-    do {
-      if (signaled) {
-        console.log(
-          "Folders selection changed, should start the sync all over again."
-        );
-        break;
+    let foldersToBrowse: string[] = await getFoldersToSync(connectorId);
+    while (foldersToBrowse.length > 0) {
+      const folderId = foldersToBrowse.pop();
```
Comment (spolu) [inline connectors/src/connectors/google_drive/temporal/workflows.ts:59]: Given the fact that there is no absolute guarantee that we have a DAG that means we may end up in an infinite loop.

I think we want to store folders that we traverse and skip the ones already synced in that run?

**Review by spolu — COMMENTED** [review]
Looks like a very reasonable starting point but I think we must store folders on top of files to avoid multi sync of the same folders and potential infinite loops given the lack of guarantees from gdrive that we have a DAG

File: connectors/src/connectors/google_drive/temporal/workflows.ts (hunk 2 above)
Comment (lasryaric) [reply connectors/src/connectors/google_drive/temporal/workflows.ts:59]: Yes I actually thought about it but I was a bit scared by Temporal having to serialize the big list of visited folders already visited each time we visit one (to save the workflow state).
I did run a little workflow test and observed no added latency even after a lot of strings being saved in memory.
```
export async function testLargeLocalState(connectorId: ModelId) {
  const queue: string[] = [];
  for (let i = 0; i < 100000; i++) {
    const start = new Date();
    for (let j = 0; j < 10000; j++) queue.push(`item-${i}-${j}`);
    await syncSucceeded(connectorId);
    const finish = new Date();
    console.log(
      `syncSucceeded took ${finish.getTime() - start.getTime()}ms - size: ${
        queue.length
      }`
    );
  }
}

```

I am pushing another commit right now.

File: connectors/src/connectors/google_drive/temporal/workflows.ts (hunk 1 above)
Comment (lasryaric) [reply connectors/src/connectors/google_drive/temporal/workflows.ts:63]: good catch. Thanks.

File: connectors/src/connectors/google_drive/temporal/workflows.ts (hunk 2 above)
Comment (spolu) [reply connectors/src/connectors/google_drive/temporal/workflows.ts:60]: I was more thinking about storing them in DB. Will be useful to show permissions as well ?

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**